diff -ur --strip-trailing-cr irrlicht/src-old/CGUIEditBox.cpp irrlicht/src/CGUIEditBox.cpp
--- irrlicht/src-old/CGUIEditBox.cpp	2016-09-15 18:12:49.532839000 +0000
+++ irrlicht/src/CGUIEditBox.cpp	2016-09-15 18:13:37.101559800 +0000
@@ -333,13 +333,7 @@
 				const c16* p = Operator->getTextFromClipboard();
 				if (p)
 				{
-					// TODO: we should have such a function in core::string
-					size_t lenOld = strlen(p);
-					wchar_t *ws = new wchar_t[lenOld + 1];
-					size_t len = mbstowcs(ws,p,lenOld);
-					ws[len] = 0;
-					irr::core::stringw widep(ws);
-					delete[] ws;
+					irr::core::stringw widep(p);
 
 					if (MarkBegin == MarkEnd)
 					{
